src/Yesod/Markdown.hs

```diff
@@ -107,25 +111,73 @@ markdownFromFile f = do
             return $ decodeUtf8With lenientDecode bs
 
 writePandoc :: WriterOptions -> Pandoc -> Html
-writePandoc wo = preEscapedToMarkup . sanitizeBalance . T.pack . writeHtmlString wo
+writePandoc wo = preEscapedToMarkup . sanitizeBalance . writeHTML wo
 
 writePandocTrusted :: WriterOptions -> Pandoc -> Html
-writePandocTrusted wo = preEscapedToMarkup . writeHtmlString wo
+writePandocTrusted wo = preEscapedToMarkup . writeHTML wo
+
+writeHTML :: WriterOptions -> Pandoc -> Text
+writeHTML wo p =
+#if MIN_VERSION_pandoc(2,0,0)
+  handleErr . runPure $ writeHtml5String wo p
+#else
+  T.pack $ writeHtmlString wo p
+#endif
 
 parseMarkdown :: ReaderOptions -> Markdown -> Either PandocError Pandoc
-parseMarkdown ro = readMarkdown ro . T.unpack . unMarkdown
+parseMarkdown ro =
+#if MIN_VERSION_pandoc(2,0,0)
+  runPure . readMarkdown ro . unMarkdown
+#else
+  readMarkdown ro . T.unpack . unMarkdown
+#endif
+
+handleErr :: Either PandocError a -> a
+handleErr =
+#if MIN_VERSION_pandoc(2,0,0)
+  unsafePerformIO . handleError
```

Comment:
I think instead of using `unsafePerformIO` we should just write and use the pandoc-1 `handleError` ourselves (which uses `error`):

```hs
-- | An unsafe method to handle `PandocError`s.
handleError :: Either PandocError a -> a
handleError (Right r) = r
handleError (Left err) =
  case err of
    ParseFailure string -> error string
    ParsecError input err' ->
        let errPos = errorPos err'
            errLine = sourceLine errPos
            errColumn = sourceColumn errPos
            ls = lines input ++ [""]
            errorInFile = if length ls > errLine - 1
                            then concat ["\n", (ls !! (errLine - 1))
                                        ,"\n", replicate (errColumn - 1) ' '
                                        ,"^"]
                        else ""
        in  error $ "\nError at " ++ show  err'
                ++ errorInFile
```

I'm not happy about it, but

1. It's the only way we can support `ToMarkup Markdown`, which is really a primary use-case
1. At least this preserves the *exact* same form of (un)safety from the versions before

WDYT?

Reply:
I can try to get something like this working.  `PandocError` from pandoc-2 has a bunch of extra error cases, so we'll have to handle those as well in our `handleError` function:

```haskell
data PandocError = PandocIOError String IOError
                 | PandocHttpError String HttpException
                 | PandocShouldNeverHappenError String
                 | PandocSomeError String
                 | PandocParseError String
                 | PandocParsecError Input ParseError
                 | PandocMakePDFError String
                 | PandocOptionError String
                 | PandocSyntaxMapError String
                 | PandocFailOnWarningError
                 | PandocPDFProgramNotFoundError String
                 | PandocPDFError String
                 | PandocFilterError String String
                 | PandocCouldNotFindDataFileError String
                 | PandocResourceNotFound String
                 | PandocTemplateError String
                 | PandocAppError String
                 | PandocEpubSubdirectoryError String
                 | PandocMacroLoop String
                 deriving (Show, Typeable, Generic)

-- | Handle PandocError by exiting with an error message.
handleError :: Either PandocError a -> IO a
handleError (Right r) = return r
handleError (Left e) =
  case e of
    PandocIOError _ err' -> ioError err'
    PandocHttpError u err' -> err 61 $
      "Could not fetch " ++ u ++ "\n" ++ show err'
    PandocShouldNeverHappenError s -> err 62 s
    PandocSomeError s -> err 63 s
    PandocParseError s -> err 64 s
    PandocParsecError input err' ->
        let errPos = errorPos err'
            errLine = sourceLine errPos
            errColumn = sourceColumn errPos
            ls = lines input ++ [""]
            errorInFile = if length ls > errLine - 1
                            then concat ["\n", ls !! (errLine - 1)
                                        ,"\n", replicate (errColumn - 1) ' '
                                        ,"^"]
                        else ""
        in  err 65 $ "\nError at " ++ show  err' ++
                     -- if error comes from a chunk or included file,
                     -- then we won't get the right text this way:
                     if sourceName errPos == "source"
                        then errorInFile
                        else ""
    PandocMakePDFError s -> err 65 s
    PandocOptionError s -> err 2 s
    PandocSyntaxMapError s -> err 67 s
    PandocFailOnWarningError -> err 3 "Failing because there were warnings."
    PandocPDFProgramNotFoundError pdfprog -> err 47 $
        pdfprog ++ " not found. Please select a different --pdf-engine or install " ++ pdfprog
    PandocPDFError logmsg -> err 43 $ "Error producing PDF.\n" ++ logmsg
    PandocFilterError filtername msg -> err 83 $ "Error running filter " ++
        filtername ++ ":\n" ++ msg
    PandocCouldNotFindDataFileError fn -> err 97 $
        "Could not find data file " ++ fn
    PandocResourceNotFound fn -> err 99 $
        "File " ++ fn ++ " not found in resource path"
    PandocTemplateError s -> err 5 s
    PandocAppError s -> err 1 s
    PandocEpubSubdirectoryError s -> err 31 $
      "EPUB subdirectory name '" ++ s ++ "' contains illegal characters"
    PandocMacroLoop s -> err 91 $
      "Loop encountered in expanding macro " ++ s

err :: Int -> String -> IO a
err exitCode msg = do
  UTF8.hPutStrLn stderr msg
  exitWith $ ExitFailure exitCode
  return undefined

```

Reply:
Gotcha. I think I'm OK with

```hs
handleError :: Either PandocError a -> a
handleError = either (error . show) id
```

Sorry if you put a lot of time into the more robust version...

Reply:
Okay, no problem.  Luckily I haven't started working on it yet :-)

Let's go with this simpler version.

Reply:
I've changed the handleError function to be as you've described above.